File: src/com/infoecos/cn/destiny/lib/支/地支.java
package com.infoecos.cn.destiny.lib.支;

import java.util.HashMap;
import java.util.Map;

import com.infoecos.cn.destiny.lib.术数;
import com.infoecos.cn.destiny.lib.common.人体;
import com.infoecos.cn.destiny.lib.common.方位;
import com.infoecos.cn.destiny.lib.common.脏腑;
import com.infoecos.cn.destiny.lib.五行.五行;
import com.infoecos.cn.destiny.lib.干.天干;


public enum 地支 {
	子, 丑, 寅, 卯, 辰, 巳, 午, 未, 申, 酉, 戌, 亥;

	private static Map<Integer, 地支> stems = new HashMap<Integer, 地支>();

	/**
	 * 返回甲子数x对应的地支
	 * 
	 * @param x
	 * @return
	 */
	public static 地支 lookup(int x) {
		x = x % 12;
		return stems.get(x < 0 ? x + 12 : x);
	}

	private 术数 shushu;
	private 天干[] 支藏;

	private String key;

	private 方位 position;
	private 人体 body;
	private 脏腑[] organs;

	private void initialize(int value, String key, 天干[] 支藏, 人体 body, 脏腑[] organs) {
		value = value % 12;

		五行 element = null;

		switch (value / 3) {
		case 0:
			element = 五行.水;
			position = 方位.北;
			break;
		case 1:
			element = 五行.木;
			position = 方位.东;
			break;
		case 2:
			element = 五行.火;
			position = 方位.南;
			break;
		case 3:
			element = 五行.金;
			position = 方位.西;
			break;
		}
		if (value % 3 == 2)
			element = 五行.土;

		this.key = key;
		this.支藏 = 支藏;
		this.body = body;
		this.organs = organs;

		shushu = new 术数(value, element);
		stems.put(value, this);
	}

	static {
		子.initialize(1, "子", new 天干[] { 天干.癸 }, 人体.耳, new 脏腑[] { 脏腑.膀胱, 脏腑.三焦 });
		丑.initialize(2, "丑", new 天干[] { 天干.己, 天干.癸, 天干.辛 }, 人体.胞肚,
				new 脏腑[] { 脏腑.脾 });
		寅.initialize(3, "寅", new 天干[] { 天干.甲, 天干.丙, 天干.戊 }, 人体.手,
				new 脏腑[] { 脏腑.胆 });
		卯.initialize(4, "卯", new 天干[] { 天干.乙 }, 人体.指, new 脏腑[] { 脏腑.肝 });
		辰.initialize(5, "辰", new 天干[] { 天干.戊, 天干.乙, 天干.癸 }, 人体.肩胸,
				new 脏腑[] { 脏腑.胃 });
		巳.initialize(6, "巳", new 天干[] { 天干.丙, 天干.戊, 天干.庚 }, 人体.面咽齿,
				new 脏腑[] { 脏腑.心 });
		午.initialize(7, "午", new 天干[] { 天干.丁, 天干.己 }, 人体.眼, new 脏腑[] { 脏腑.小肠 });
		未.initialize(8, "未", new 天干[] { 天干.己, 天干.丁, 天干.乙 }, 人体.脊梁,
				new 脏腑[] { 脏腑.脾 });
		申.initialize(9, "申", new 天干[] { 天干.庚, 天干.壬, 天干.戊 }, 人体.经络,
				new 脏腑[] { 脏腑.大肠 });
		酉.initialize(10, "酉", new 天干[] { 天干.辛 }, 人体.精血, new 脏腑[] { 脏腑.肺 });
		戌.initialize(11, "戌", new 天干[] { 天干.戊, 天干.辛, 天干.丁 }, 人体.命门腿足,
				new 脏腑[] { 脏腑.胃 });
		亥.initialize(12, "亥", new 天干[] { 天干.壬, 天干.甲 }, 人体.头, new 脏腑[] { 脏腑.肾,
				脏腑.心包 });
	}

	public 术数 术数() {
		return shushu;
	}

	public 方位 方位() {
		return position;
	}

	public 人体 人体() {
		return body;
	}

	public 脏腑[] 脏腑() {
		return organs;
	}

	public 天干[] 支藏() {
		return 支藏;
	}

	@Override
	public String toString() {
		return key;
	}

	public 五行 六合(地支 x) {
		if ((子.equals(x) && 丑.equals(this)) || (子.equals(this) && 丑.equals(x)))
			return 五行.土;
		if ((寅.equals(x) && 亥.equals(this)) || (寅.equals(this) && 亥.equals(x)))
			return 五行.木;
		if ((卯.equals(x) && 戌.equals(this)) || (卯.equals(this) && 戌.equals(x)))
			return 五行.火;
		if ((辰.equals(x) && 酉.equals(this)) || (辰.equals(this) && 酉.equals(x)))
			return 五行.金;
		if ((巳.equals(x) && 申.equals(this)) || (巳.equals(this) && 申.equals(x)))
			return 五行.水;
		if ((午.equals(x) && 未.equals(this)) || (午.equals(this) && 未.equals(x)))
			return 五行.土;

		return null;
	}

	/**
	 * 会一方之气。
	 * 
	 * @param x
	 * @param y
	 * @param z
	 * @return
	 */
	public static 五行 三会(地支 x, 地支 y, 地支 z) {
		if (亥.equals(x) || 亥.equals(y) || 亥.equals(z)) {
			if (子.equals(x) || 子.equals(y) || 子.equals(z)) {
				if (丑.equals(x) || 丑.equals(y) || 丑.equals(z))
					return 五行.水;
			}
		} else if (寅.equals(x) || 寅.equals(y) || 寅.equals(z)) {
			if (卯.equals(x) || 卯.equals(y) || 卯.equals(z)) {
				if (辰.equals(x) || 辰.equals(y) || 辰.equals(z))
					return 五行.木;
			}
		} else if (巳.equals(x) || 巳.equals(y) || 巳.equals(z)) {
			if (午.equals(x) || 午.equals(y) || 午.equals(z)) {
				if (未.equals(x) || 未.equals(y) || 未.equals(z))
					return 五行.火;
			}
		} else if (申.equals(x) || 申.equals(y) || 申.equals(z)) {
			if (酉.equals(x) || 酉.equals(y) || 酉.equals(z)) {
				if (戌.equals(x) || 戌.equals(y) || 戌.equals(z))
					return 五行.金;
			}
		}

		return null;
	}

	/**
	 * 生、旺、墓成局。三合局力量次于三会局。
	 * 
	 * 例，水生于申，旺于子，墓于辰，故申子辰合水局
	 * 
	 * 丑辰未戌全，方是土局
	 * 
	 * @param x
	 * @param y
	 * @param z
	 * @return
	 */
	public static 五行 三合(地支 x, 地支 y, 地支 z) {
		if (申.equals(x) || 申.equals(y) || 申.equals(z)) {
			if (子.equals(x) || 子.equals(y) || 子.equals(z)) {
				if (辰.equals(x) || 辰.equals(y) || 辰.equals(z))
					return 五行.水;
			}
		} else if (巳.equals(x) || 巳.equals(y) || 巳.equals(z)) {
			if (酉.equals(x) || 酉.equals(y) || 酉.equals(z)) {
				if (丑.equals(x) || 丑.equals(y) || 丑.equals(z))
					return 五行.金;
			}
		} else if (寅.equals(x) || 寅.equals(y) || 寅.equals(z)) {
			if (午.equals(x) || 午.equals(y) || 午.equals(z)) {
				if (戌.equals(x) || 戌.equals(y) || 戌.equals(z))
					return 五行.火;
			}
		} else if (亥.equals(x) || 亥.equals(y) || 亥.equals(z)) {
			if (卯.equals(x) || 卯.equals(y) || 卯.equals(z)) {
				if (未.equals(x) || 未.equals(y) || 未.equals(z))
					return 五行.木;
			}
		}

		return null;
	}

	/**
	 * 三合半合局，力量次于三合局
	 * 
	 * @param x
	 * @param y
	 * @return
	 */
	public static 五行 生地半合(地支 x, 地支 y) {
		if ((卯.equals(x) && 亥.equals(y)) || (卯.equals(y) && 亥.equals(x)))
			return 五行.木;
		if ((巳.equals(x) && 酉.equals(y)) || (巳.equals(y) && 酉.equals(x)))
			return 五行.金;
		if ((子.equals(x) && 申.equals(y)) || (子.equals(y) && 申.equals(x)))
			return 五行.水;
		if ((午.equals(x) && 寅.equals(y)) || (午.equals(y) && 寅.equals(x)))
			return 五行.火;
		return null;
	}

	/**
	 * 三合半合局，力量次于生地半合
	 * 
	 * @param x
	 * @param y
	 * @return
	 */
	public static 五行 墓地半合(地支 x, 地支 y) {
		if ((未.equals(x) && 卯.equals(y)) || (未.equals(y) && 卯.equals(x)))
			return 五行.木;
		if ((丑.equals(x) && 酉.equals(y)) || (丑.equals(y) && 酉.equals(x)))
			return 五行.金;
		if ((子.equals(x) && 辰.equals(y)) || (子.equals(y) && 辰.equals(x)))
			return 五行.水;
		if ((午.equals(x) && 戌.equals(y)) || (午.equals(y) && 戌.equals(x)))
			return 五行.火;
		return null;
	}

	/**
	 * 三合半合局，力量次于生地半合
	 * 
	 * @param x
	 * @param y
	 * @return
	 */
	public static 五行 非旺半合(地支 x, 地支 y) {
		if ((未.equals(x) && 亥.equals(y)) || (未.equals(y) && 亥.equals(x)))
			return 五行.木;
		if ((丑.equals(x) && 巳.equals(y)) || (丑.equals(y) && 巳.equals(x)))
			return 五行.金;
		if ((申.equals(x) && 辰.equals(y)) || (申.equals(y) && 辰.equals(x)))
			return 五行.水;
		if ((寅.equals(x) && 戌.equals(y)) || (寅.equals(y) && 戌.equals(x)))
			return 五行.火;

		return null;
	}
}

File: libs/destiny/支/地支.php
<?php
namespace DchLib;

class 地支 {
	public static $子, $丑, $寅, $卯, $辰, $巳, $午, $未, $申, $酉, $戌, $亥;
	private static $values = array();

	/**
	 * 返回甲子数x对应的地支
	 * 
	 * @param $x
	 * @return
	 */
	public static function lookup($x) {
		$x = $x % 12;
		return self::$values[($x < 0 ? $x + 12 : $x)];
	}

	public static function __init() {
		self::$子 = new 地支(1, '子', array( 天干::$癸 ), 人体::$耳, array( 脏腑::$膀胱, 脏腑::$三焦 ));
		self::$丑 = new 地支(2, '丑', array( 天干::$己, 天干::$癸, 天干::$辛 ), 人体::$胞肚, array( 脏腑::$脾 ));
		self::$寅 = new 地支(3, '寅', array( 天干::$甲, 天干::$丙, 天干::$戊 ), 人体::$手, array( 脏腑::$胆 ));
		self::$卯 = new 地支(4, '卯', array( 天干::$乙 ), 人体::$指, array( 脏腑::$肝 ));
		self::$辰 = new 地支(5, '辰', array( 天干::$戊, 天干::$乙, 天干::$癸 ), 人体::$肩胸, array( 脏腑::$胃 ));
		self::$巳 = new 地支(6, '巳', array( 天干::$丙, 天干::$戊, 天干::$庚 ), 人体::$面咽齿, array( 脏腑::$心 ));
		self::$午 = new 地支(7, '午', array( 天干::$丁, 天干::$己 ), 人体::$眼, array( 脏腑::$小肠 ));
		self::$未 = new 地支(8, '未', array( 天干::$己, 天干::$丁, 天干::$乙 ), 人体::$脊梁, array( 脏腑::$脾 ));
		self::$申 = new 地支(9, '申', array( 天干::$庚, 天干::$壬, 天干::$戊 ), 人体::$经络, array( 脏腑::$大肠 ));
		self::$酉 = new 地支(10, '酉', array( 天干::$辛 ), 人体::$精血, array( 脏腑::$肺 ));
		self::$戌 = new 地支(11, '戌', array( 天干::$戊, 天干::$辛, 天干::$丁 ), 人体::$命门腿足, array( 脏腑::$胃 ));
		self::$亥 = new 地支(12, '亥', array( 天干::$壬, 天干::$甲 ), 人体::$头, array( 脏腑::$肾, 脏腑::$心包 ));

		self::$values = array(
			self::$亥, self::$子, self::$丑, self::$寅, self::$卯, self::$辰, self::$巳, self::$午, self::$未, self::$申, self::$酉, self::$戌
		);
	}
	
	private $m_shushu;
	private $支藏;

	private $m_key;

	private $m_position;
	private $m_body;
	private $m_organs;

	private function __construct($value, $key, $支藏, $body, $organs) {
		$value = $value % 12;

		switch ($value / 3) {
		case 0:
			$element = 五行::$水;
			$position = 方位::$北;
			break;
		case 1:
			$element = 五行::$木;
			$position = 方位::$东;
			break;
		case 2:
			$element = 五行::$火;
			$position = 方位::$南;
			break;
		case 3:
			$element = 五行::$金;
			$position = 方位::$西;
			break;
		}
		if ($value % 3 == 2)
			$element = 五行::$土;

		$this->m_key = $key;
		$this->m_position = $position;
		$this->支藏 = $支藏;
		$this->m_body = $body;
		$this->m_organs = $organs;

		$this->m_shushu = new 术数($element, null, $value);
	}

	public function 术数() {
		return $this->m_shushu;
	}

	public function 方位() {
		return $this->m_position;
	}

	public function 人体() {
		return $this->m_body;
	}

	public function 脏腑() {
		return $this->m_organs;
	}

	public function 支藏() {
		return $this->支藏;
	}

	public function toString() {
		return $this->m_key;
	}

	public function 六合($x) {
		if ((self::$子 === $x && self::$丑 === $this) || (self::$子 === $this && self::$丑 === $x))
			return 五行::$土;
		if ((self::$寅 === $x && self::$亥 === $this) || (self::$寅 === $this && self::$亥 === $x))
			return 五行::$木;
		if ((self::$卯 === $x && self::$戌 === $this) || (self::$卯 === $this && self::$戌 === $x))
			return 五行::$火;
		if ((self::$辰 === $x && self::$酉 === $this) || (self::$辰 === $this && self::$酉 === $x))
			return 五行::$金;
		if ((self::$巳 === $x && self::$申 === $this) || (self::$巳 === $this && self::$申 === $x))
			return 五行::$水;
		if ((self::$午 === $x && self::$未 === $this) || (self::$午 === $this && self::$未 === $x))
			return 五行::$土;

		return null;
	}

	/**
	 * 会一方之气。
	 * 
	 * @param $x
	 * @param $y
	 * @param $z
	 * @return
	 */
	public static function 三会($x, $y, $z) {
		if (self::$亥 === $x || self::$亥 === $y || self::$亥 === $z) {
			if (self::$子 === $x || self::$子 === $y || self::$子 === $z) {
				if (self::$丑 === $x || self::$丑 === $y || self::$丑 === $z)
					return 五行::$水;
			}
		} else if (self::$寅 === $x || self::$寅 === $y || self::$寅 === $z) {
			if (self::$卯 === $x || self::$卯 === $y || self::$卯 === $z) {
				if (self::$辰 === $x || self::$辰 === $y || self::$辰 === $z)
					return 五行::$木;
			}
		} else if (self::$巳 === $x || self::$巳 === $y || self::$巳 === $z) {
			if (self::$午 === $x || self::$午 === $y || self::$午 === $z) {
				if (self::$未 === $x || self::$未 === $y || self::$未 === $z)
					return 五行::$火;
			}
		} else if (self::$申 === $x || self::$申 === $y || self::$申 === $z) {
			if (self::$酉 === $x || self::$酉 === $y || self::$酉 === $z) {
				if (self::$戌 === $x || self::$戌 === $y || self::$戌 === $z)
					return 五行::$金;
			}
		}

		return null;
	}

	/**
	 * 生、旺、墓成局。三合局力量次于三会局。
	 * 
	 * 例，水生于申，旺于子，墓于辰，故申子辰合水局
	 * 
	 * 丑辰未戌全，方是土局
	 * 
	 * @param $x
	 * @param $y
	 * @param $z
	 * @return
	 */
	public static function 三合($x, $y, $z) {
		if (self::$申 === $x || self::$申 === $y || self::$申 === $z) {
			if (self::$子 === $x || self::$子 === $y || self::$子 === $z) {
				if (self::$辰 === $x || self::$辰 === $y || self::$辰 === $z)
					return 五行::$水;
			}
		} else if (self::$巳 === $x || self::$巳 === $y || self::$巳 === $z) {
			if (self::$酉 === $x || self::$酉 === $y || self::$酉 === $z) {
				if (self::$丑 === $x || self::$丑 === $y || self::$丑 === $z)
					return 五行::$金;
			}
		} else if (self::$寅 === $x || self::$寅 === $y || self::$寅 === $z) {
			if (self::$午 === $x || self::$午 === $y || self::$午 === $z) {
				if (self::$戌 === $x || self::$戌 === $y || self::$戌 === $z)
					return 五行::$火;
			}
		} else if (self::$亥 === $x || self::$亥 === $y || self::$亥 === $z) {
			if (self::$卯 === $x || self::$卯 === $y || self::$卯 === $z) {
				if (self::$未 === $x || self::$未 === $y || self::$未 === $z)
					return 五行::$木;
			}
		}

		return null;
	}

	/**
	 * 三合半合局，力量次于三合局
	 * 
	 * @param $x
	 * @param $y
	 * @return
	 */
	public static function 生地半合($x, $y) {
		if ((self::$卯 === $x && self::$亥 === $y) || (self::$卯 === $y && self::$亥 === $x))
			return 五行::$木;
		if ((self::$巳 === $x && self::$酉 === $y) || (self::$巳 === $y && self::$酉 === $x))
			return 五行::$金;
		if ((self::$子 === $x && self::$申 === $y) || (self::$子 === $y && self::$申 === $x))
			return 五行::$水;
		if ((self::$午 === $x && self::$寅 === $y) || (self::$午 === $y && self::$寅 === $x))
			return 五行::$火;
		return null;
	}

	/**
	 * 三合半合局，力量次于生地半合
	 * 
	 * @param $x
	 * @param $y
	 * @return
	 */
	public static function 墓地半合($x, $y) {
		if ((self::$未 === $x && self::$卯 === $y) || (self::$未 === $y && self::$卯 === $x))
			return 五行::$木;
		if ((self::$丑 === $x && self::$酉 === $y) || (self::$丑 === $y && self::$酉 === $x))
			return 五行::$金;
		if ((self::$子 === $x && self::$辰 === $y) || (self::$子 === $y && self::$辰 === $x))
			return 五行::$水;
		if ((self::$午 === $x && self::$戌 === $y) || (self::$午 === $y && self::$戌 === $x))
			return 五行::$火;
		return null;
	}

	/**
	 * 三合半合局，力量次于生地半合
	 * 
	 * @param $x
	 * @param $y
	 * @return
	 */
	public static function 非旺半合($x, $y) {
		if ((self::$未 === $x && self::$亥 === $y) || (self::$未 === $y && self::$亥 === $x))
			return 五行::$木;
		if ((self::$丑 === $x && self::$巳 === $y) || (self::$丑 === $y && self::$巳 === $x))
			return 五行::$金;
		if ((self::$申 === $x && self::$辰 === $y) || (self::$申 === $y && self::$辰 === $x))
			return 五行::$水;
		if ((self::$寅 === $x && self::$戌 === $y) || (self::$寅 === $y && self::$戌 === $x))
			return 五行::$火;

		return null;
	}
}

地支::__init();